feat(gatsby-source-wordpress): add support for route whitelist to complement route blacklist option

lightstrike opened the pull request:
First shot at #5308 -- I need to actually test this (need to get on the v2 train), curious for thoughts on developer experience (DX) and whether default whitelist (better for beginners and probably most use cases as @KyleAMathews [implies](https://github.com/gatsbyjs/gatsby/issues/5308#issuecomment-387078677)) or default blacklist (preserves backwards compatibility) is better.

m-allanson commented:
Thanks for taking this on! Like you say there are a couple of approaches that could be used here.

Looking at your current changes, I think the `routeListType` option is quite confusing and could probably be dropped.

The behaviour could be inferred based on which of `includedRoutes` or `excludedRoutes` exists:

- `includedRoutes` only, use whitelist behaviour
- `excludedRoutes` only, use blacklist behaviour
- both exist - use whitelist behaviour and optionally show a warning and / or merge the blacklist into the whitelist? So you could whitelist `["posts/**"]` and then blacklist `["posts/dont-include-me"]`

I'm not sure what the default behaviour should be if neither option exists. The plugin is getting a major version bump for Gatsby v2, so that would be an appropriate time to make a breaking change if necessary. If this is going to be a breaking change then we should also add a deprecation notice to v1 of the plugin.

pieh commented:
> both exist - use whitelist behaviour and optionally show a warning and / or merge the blacklist into the whitelist? So you could whitelist ["posts/**"] and then blacklist ["posts/dont-include-me"]

I would apply whitelist first and blacklist after that, when determining if route should be fetched.

> I'm not sure what the default behaviour should be if neither option exists. The plugin is getting a major version bump for Gatsby v2, so that would be an appropriate time to make a breaking change if necessary. 

I would set default `includedRoutes` to handle default vanilla wordpress installation - so include posts, pages, categories, tags, media and authors. I would say that we can use v2 major version bump to justify changing it to include fewer routes by default and put information about this change in prominent spot in docs with example config that would fetch all routes as it does now (if someone is relying on that). This would probably also involve wordpress guides changes to cover that.

lightstrike commented:
Thanks for the feedback! I agree whitelist default is the best path forward. I've updated to reflect this approach and also removed `routeListType`.

How do you both feel about removing `excludedTypes`? I'm not sure what the story behind their inclusion is.

lightstrike commented:
I just cleaned up the whitelist/blacklist logic and tested it with a sample project. It think it works now, would definitely be great for another person to check.

pieh commented on packages/gatsby-source-wordpress/src/fetch.js:
```diff
@@ -453,6 +466,11 @@ function getValidRoutes({
+        if (_verbose)
+          console.log(
+            colorized.out(`Invalid route.`, colorized.color.Font.FgRed)
+          )
```
Maybe we could keep track why this route is excluded and print out that either this route wasn't whitelisted or it was blacklisted?

lightstrike commented:
Added some messaging to that effect!

Also removed `baseUrl` as I don't see a use case for it here. Let me know if I'm missing something.

lightstrike commented:
Making a note here, let's include this resource in the docs somewhere as it's really helpful: http://pthrasher.github.io/minimatch-test/

lightstrike commented:
@m-allanson @pieh Anything else you'd like me to consider here?

pieh commented on packages/gatsby-source-wordpress/src/fetch.js:
```diff
@@ -355,21 +356,40 @@ async function getPages(
+  } else {
+    return false
+  }
+}
```
`Array.prototype.some` returns `true` or `false` so this could be shorter:
```js
function checkRouteList( routePath, routeList ) {
  return routeList.some(route =>
    minimatch(routePath, route)
  )
}
```

lightstrike replied:
Good catch :smiley:

lightstrike commented:
@pieh Ran the linting fixes and added some docs.

> add changelog to `README.md`

What do you mean by this?

Also do you want me to rebase with the latest master commit?

KyleAMathews commented:
Deploy preview for *using-postcss-sass* failed.

Built with commit 7e0c2a1efbc8d0a607aa8371727ba836d45aef94

https://app.netlify.com/sites/using-postcss-sass/deploys/5b50e32bb3127458044ae5f7

pieh commented:
> add changelog to README.md
> 
> What do you mean by this?

Because we won't be pulling from all REST endpoint, we need to add information to https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-wordpress/README.md that documents this change and provide quick guide how to restore previous behaviour (I think using `includedRoutes: ['*/*/*']` would allow to fetch all available endpoints) and if possible link to description of `includedRoutes` option on how to have more granular control over fetched routes

KyleAMathews commented:
Deploy preview for *gatsbyjs* failed.

Built with commit 958f4586ba24d0c94fcf009471c7615d9b9b3661

https://app.netlify.com/sites/gatsbyjs/deploys/5b5121c067610c5af1a0ec53

lightstrike commented:
@pieh I made those changes. Let me know what you think.

Should we update [the example](https://github.com/gatsbyjs/gatsby/tree/master/examples/using-wordpress) and [tutorial docs](https://github.com/gatsbyjs/gatsby/blob/master/docs/docs/wordpress-source-plugin-tutorial.md) as part of this PR as well?

lightstrike commented:
Also another thought: while it's certainly widely used, can we remove the special treatment of `Advanced Custom Fields`? Those wanting to use it could add `/acf/v2/**`, `/acf/v3/**` or whatever permutation they wanted. Tangentially related to #5796 (I have some specific thoughts on this I'll share soon).

We could take a soft deprecation approach at first as it's likely to break a lot of builds if we fully removed the `useACF` option right away.

pieh commented:
I think using `useACF` just adds `options` page to routeList because it's not declared in root endpoint (not sure here)?
https://github.com/gatsbyjs/gatsby/blob/518a5829f77a656464778fe578088b9503ce3392/packages/gatsby-source-wordpress/src/fetch.js#L462-L482

So maybe this should be other way around - if someone set `useACF` to true we should add `/acf/*/*` to whitelist?

pieh commented on packages/gatsby-source-wordpress/README.md:
```diff
@@ -106,6 +121,33 @@ plugins: [
+
+To customize the routes fetched, two configuration options are available: `includeRoutes` for whitelisting and `excludeRoutes` for blacklisting. Both options expect an array of glob patterns. Glob matching is done by [minimatch](https://github.com/isaacs/minimatch). To test your glob patterns, [use this tool](http://pthrasher.github.io/minimatch-test/).
+
+If there are
```
Not finished sentence :)

lightstrike replied:
Oops, not needed. Will remove.

pieh commented on packages/gatsby-source-wordpress/README.md:
```diff
@@ -106,6 +121,33 @@ plugins: [
+
+By default, it will ignore all other endpoints listed in the root `/wp-json` endpoint.
+
+To customize the routes fetched, two configuration options are available: `includeRoutes` for whitelisting and `excludeRoutes` for blacklisting. Both options expect an array of glob patterns. Glob matching is done by [minimatch](https://github.com/isaacs/minimatch). To test your glob patterns, [use this tool](http://pthrasher.github.io/minimatch-test/).
```
I don't remember right now - do we print out urls without `verbose` option?

lightstrike replied:
No, only the type and how many entities [were fetched](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-wordpress/src/fetch.js#L257).

lightstrike commented:
@pieh Started to rebase with master given the conflict and had a question.

What is [this line about](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-wordpress/src/fetch.js#L413)?

```js
// Grab ACF Version from routes
acfRestVersion = key === `/acf/${entityType}` ? entityType.substr(1) : acfRestVersion
```

Should it be inside [this verbose block](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-wordpress/src/fetch.js#L475) instead?

Khristophor commented:
@lightstrike I'll try to fill in some details here based on my (and I assume others') use case with ACF Options and source-wordpress.

The ACF to REST plugin has two versions that are selectable from the WordPress admin area, v2 and v3. Everything seems to be fairly similar with how the data is returned through the REST API with the exception of the [ACF Options page(s)](https://www.advancedcustomfields.com/add-ons/options-page/). The v2 options page is found at `/acf/v2/options`, while the v3 options page can be found at `/acf/v3/options/options` _or_ `/acf/v3/options/[option page ID]` if custom options page ID's are used.

The ACF to REST plugin does not provide these routes through a `_links` property in the REST API similar to other resources, so Gatsby is unable to query them automatically.

I think including the logic surrounding the ACF REST version from #6556 as well as whitelisting `/acf/*/*` if useACF is found as @pieh suggested is the most painless migration path with your proposed changes.

lightstrike commented:
@Khristophor thanks for the comments. For now I agree with this sentiment:

> I think including the logic surrounding the ACF REST version from #6556 as well as whitelisting /acf/*/* if useACF is found as @pieh suggested is the most painless migration path with your proposed changes.

I was thinking about generalized support for detail routes and think a `detailedRoutes` configuration option could do well, along the lines of:

```js
detailRoutes: [
  {
    endpoint: `/acf/v3/options/`,
    resources: [`option_page_1`, `option_page_2`],
  },
]
```

I'd like to support the source plugin being a solid generalized base, with the goal to provide further extensibility along the lines expressed in #5796. In ACF's case, I believe the combination of `includedRoutes` and `detailRoutes` could eliminate the need for `useACF` and `acfOptionPageIds` within the core plugin itself, where a `gatsby-wordpress-acf` sub-plugin could provide extra features like the data combination in #6556. FWIW, the `combineACF` function could be achieved through the current custom normalizer hook.

I'm also thinking through how to streamline the data transformation pipeline currently done in `normalize.js` - would be great to get your thoughts once I have those together.

lightstrike commented:
I'm not sure why it's needed at all as we have the `acfRestVersion` config option -- @DanArpoika can you clarify the purpose of [this line](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-wordpress/src/fetch.js#L413)?

pieh commented:
> I'm not sure why it's needed at all as we have the acfRestVersion config option

We actually don't have it - it was there initially, but because we can extract version from endpoints urls, @DanArpoika changed PR to do automatic extraction.

Khristophor commented:
@lightstrike we wanted to avoid any extra margin for error around setting a config option in Gatsby that could be different from the setting in WordPress.

Now Gatsby will simply detect the ACF to REST version being used, no configuration setting needed.

DanArpoika commented:
@lightstrike Sorry, little late to the party as I've been traveling.

My original plan was to set the ACF to REST version manually in the config, but as @pieh mentioned it is easier and leaves less room for error detecting it automatically from the endpoints.

lightstrike commented:
@pieh @Khristophor @DanArpoika Thanks all for the feedback!

Finally got the chance to take a look at this again with fresh eyes. I pushed an update that puts all the ACF logic into one conditional block based on `useACF`, which will help with any refactoring of the ACF logic into its own sub-plugin in the future.

I also rewrote the version detection logic to be driven off the `namespaces` property in the root `wp-json` endpoint, which I believe is more efficient and straightforward.

One thing I want to highlight that might result in undesired behavior:

![screenshot from 2018-08-19 13-24-39](https://user-images.githubusercontent.com/1368742/44312849-7871d680-a3b3-11e8-843e-1ab8d1f4b050.png)

This is the fetch logic based on the default `includedRoutes` along with `useACF` set to `false`. We can make it explicitly for `wp/v2` if this could cause confusion.

Please take a look when you're able.

pieh commented:
I jest tested it locally and works as expected - but I will have to do small changes here. Because I dropped the ball with merging this before major version bump, I will change default `includedRoutes` to match all routes and update docs about it. We need to remember about change that back before next major version bump

pieh commented:
@lightstrike Was just waiting to see if you want to change anything before merging - this is good as-is so let's merge it (finally!)
